// @sveltejs/kit/dist/chunks/index3.js
import fs__default from 'fs';
import path__default from 'path';

/** @param {string} dir */
function mkdirp(dir) {
	try {
		fs__default.mkdirSync(dir, { recursive: true });
	} catch (e) {
		if (e.code === 'EEXIST') return;
		throw e;
	}
}

/** @param {string} path */
function rimraf(path) {
	(fs__default.rmSync || fs__default.rmdirSync)(path, { recursive: true, force: true });
}

/**
 * @param {string} from
 * @param {string} to
 * @param {(basename: string) => boolean} filter
 */
function copy(from, to, filter = () => true) {
	if (!fs__default.existsSync(from)) return [];
	if (!filter(path__default.basename(from))) return [];

	const files = [];
	const stats = fs__default.statSync(from);

	if (stats.isDirectory()) {
		fs__default.readdirSync(from).forEach((file) => {
			files.push(...copy(path__default.join(from, file), path__default.join(to, file)));
		});
	} else {
		mkdirp(path__default.dirname(to));
		fs__default.copyFileSync(from, to);
		files.push(to);
	}

	return files;
}

export { copy as c, mkdirp as m, rimraf as r };
